File: other_algos/breadth_first_search.js
var Tree = function(value) {
  this.val = value;
  this.children = [];
}
Tree.prototype.addChild = function(val) {
  this.children.push(new Tree(val));
}
Tree.prototype.contains = function(val) {
  if (this.val === val) {
    return true;
  } else {
    for (var i = 0; i < this.children.length; i++) {
      if (this.children[i].contains(val)) {
        return true;
      }
    }
    return false;
  }
  return false;
}

//Implement bread-first search on a basic Tree using a queue
Tree.prototype.bfSearch = function(value) {
  var queue = [this];

  while (true) {
    if (queue.length === 0) {
      return false;
    } else {
      var node = queue.pop();
    }
    if (node.val === value) {
      return true;
    } else {
      for (var i = 0; i < node.children.length; i++) {
        queue.unshift(node.children[i]);
      }
    }
  }
}

File: other_algos/breadth_first_search.py
#Breadth first search of a tree.
#Tree class
class Tree():
    def __init__(self, value):
        self.val = value
        self.children = []

    def addChild(self, value):
        self.children.append(Tree(value))

    def contains(self, value):
        if self.val == value:
            return True
        else:
            for node in self.children:
                if node.contains(value):
                    return True
            return False
        
#Breadth first search
def bfs(tree, value):
    queue = [tree]
    
    while len(queue):
        node = queue.pop()
        if node.val == value:
            return True
        else:
            for child in node.children:
                queue.insert(0, child)
            return False
            
	return False

